src/math/exp_fcn.c

```diff
@@ -0,0 +1,204 @@
+		384307168202282325LL,
+		76861433640456465LL,
+		12810238940076077LL,
+		1830034134296582LL,
+		228754266787072LL,
+		25417140754119LL,
+		2541714075411LL,
+		231064915946LL,
+		19255409662LL,
+		1481185358LL,
+		105798954LL,
+		7053264LL,
+		440829LL,
+		25931LL,
+		1441LL,
+		76LL,
+		4LL
+};
+
+/* Function Declarations */
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *ptroutbitshi
+ *		  uint64_t *ptroutbitslo
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ * Description:Perform element-wise multiplication on in_0 and in_1
+ * while keeping the required product word length and fractional
+ * length in mind. mul_s64 function divide the 64-bit quantities
+ * into two 32-bit words,multiply the low words to produce the
+ * lowest and second-lowest words in the result, then both pairs
+ * of low and high words from different numbers to produce the
+ * second and third lowest words in the result, and finally both
+ * high words to produce the two highest words in the outcome.
+ * Add them all up, taking carry into consideration.
+ *
+ * The 64 x 64 bit multiplication of operands in_0 and in_1 is
+ * shown in the image below. The 64-bit operand in_0,in_1 is
+ * represented by the notation in0_H, in1_H for the top 32 bits
+ * and in0_L, in1_L for the bottom 32 bits.
+ *
+ *				in0_H : in0_L
+ *			x	in1_H : in1_L
+ *			---------------------
+ *	P0			in0_L x in1_L
+ *	P1		in0_H x in1_L		64 bit inner multiplication
+ *	P2		in0_L x in1_H		64 bit inner multiplication
+ *	P3	in0_H x in1_H
+ *			--------------------
+ *			[64 x 64 bit multiplication] sum of inner products
+ * All combinations are multiplied by one another and then added.
+ * Each inner product is moved into its proper power location.given the names
+ * of the inner products, redoing the addition where 000 represents 32 zero
+ * bits.The inner products can be added together in 64 bit addition.The sum
+ * of two 64-bit numbers yields a 65-bit output.
+ *		   (P0H:P0L)
+ *		P1H(P1L:000)
+ *		P2H(P2L:000)
+ *	P3H:P3L(000:000)
+ *	.......(aaa:P0L)
+ * By combining P0H:P0L and P1L:000. This can lead to a carry, denote as CRY0.
+ * The partial result is then multiplied by P2L:000.
+ * We call it CRY1 because it has the potential to carry again.
+ *	(CRY0 + CRY1)P0H:P0L
+ *	(	 P1H)P1L:000
+ *	(	 P2H)P2L:000
+ *	(P3H:	 P3L)000:000
+ *	--------------------
+ *	(ccc:bbb)aaa:P0L
+ * P1H, P2H, and P3H:P3L are added to the carry CRY0 + CRY1.This increase will
+ * not result in an overflow.
+ *
+ */
+static inline void mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
+{
+	uint64_t absin0 = ABS(in_0);
+	uint64_t absin1 = ABS(in_1);
+	uint64_t in0hi = absin0 >> 32;
+	uint64_t in0lo = absin0 & UINT32_MAX;
+	uint64_t in1hi = absin1 >> 32;
+	uint64_t prodhilo;
+	uint64_t prodlohi;
+
+	absin0 = absin1 & UINT32_MAX;
+	// multiplication
+	prodhilo = in0hi * absin0;
+	prodlohi = in0lo * in1hi;
+	absin0 *= in0lo;
+	absin1 = 0;
```

Comment:
I note that accuracy is compromised while absin1 is failing to reset `0x4000000000000000`. 

 ```
     index|   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|
---------+---------------+---------------+---------------+---------------+---------------+
         1|-4.921259842813|    11.02010691|     0.00728994|    20.83796843|110128.16973544|

```

Reply:
Are you saying this pointless `absin1 = 0` line makes a difference? Can you share reproduction steps?

I compiled this file with and without this line commented out and the generated `exp_fnc.c.obj` files are identical. I tried both XCC and the Zephyr SDK 0.15.2

Reply:
I 'm  saying this pointless absin1 = 0 line makes a difference.  To reproduce , you can run the  `CI Unit test `. I 've  included step by step explanation here [https://github.com/thesofproject/sof/pull/6655#discussion_r1052030947](url)

Reply:
@ShriramShastry this could be a compiler error when trying to optimize the code since we all expect this assignment to be ignored (since its assigned again later on). You and @marc-hb could have different compiler versions and ```-O``` optimization flags. Also which platform are you testing on as the older platforms have older XCC versions and different ISAs.

Reply:
> I 'm saying this pointless absin1 = 0 line makes a difference.

This looks like a very serious compiler bug! Or relying on undefined behavior.

> You and @marc-hb could have different compiler versions and -O optimization flags.

I merely compiled with the default `sof/scripts/xtensa-build-zephyr.py` script and default flags and compared the generated `exp_fnc.c.obj` files. Unlike unit tests this is what goes into the product. I made sure to keep the number of lines the same to avoid differences in debug sections (disabling all `-g` flags in Zephyr is doable but a bit tricky)

It's obvious that the optimizer will remove pointless instructions, that's the most obvious optimization. Unit tests likely have -O disabled so they can be run with a debugger but this should _still_ not change the results. So this is either a compiler bug or this code relies on some undefined behavior. Both are really bad and should be investigated.

Which toolchain(s) did you use?

> I 've included step by step explanation here https://github.com/thesofproject/sof/pull/6655#discussion_r1052030947

I could not find reproduction steps there sorry.

Reply:
You might just need to change the line to represent longs, `absin1 = in0lo < absin0 ? 1L : 0L;` cause it might be treating it like a u32 and not touching the upper 32 bits. @ShriramShastry is this the cadence compiler or gcc you are seeing this on?

Reply:
@ShriramShastry Are you using RI-2020.5? It's the tools version for MTL.  I wonder if some very new compiler version could have a bug.

Reply:
Please observe a case where an input is supplied with '0.050505049527'. I've removed 'absin1 = 0;'.  If this is acceptable, 
I tried '0.050505049527' on all compilers, including HiFi3 and HiFi4, and I'm presenting my findings. 

```
FAIL Case
               absIn0|                absIn1|                 in0Hi|                 in0Lo|                 in1Hi|                  in_0|                  in_1|          ptroutbitshi|          ptroutbitslo|
---------------------|----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+
     40000000000000ce|      4000000000000000|                     0|      c000000000000000|              40000000|                   33b|      4000000000000000|      40000000000000ce|      c000000000000000|
               cede62|                cede62|                     0|             29c486696|                     0|                   33b|                cede62|                cede62|             29c486696|
     433b79880000000a|      4000000000000000|               cede620|      8000000000000000|              40000000|       cede6200000002a|      4000000000000000|      433b79880000000a|      8000000000000000|
               d9510c|                cede62|               cede620|      6e61d84021f07c14|                     0|       cede6200000002a|                cede62|                d9510c|      6e61d84021f07c14|
     167716bb3dd7caff|      1555555555555555|               d9510c6|      d0ce4fbdb34b7eaa|              15555555|       d9510c6e61d8402|      1555555555555555|      167716bb3dd7caff|      d0ce4fbdb34b7eaa|
               d9d821|                cede62|               d9510c6|      f7aa1371e96244c4|                     0|       d9510c6e61d8402|                cede62|                d9d821|      f7aa1371e96244c4|
      59df2b5fd38b125|       555555555555555|               d9d821f|      a0cb7f4ad71f984b|               5555555|       d9d821f7aa1371f|       555555555555555|       59df2b5fd38b125|      a0cb7f4ad71f984b|
               d9def4|                cede62|               d9d821f|      8277ca1f1a75fbde|                     0|       d9d821f7aa1371f|                cede62|                d9def4|      8277ca1f1a75fbde|
      11f9765a24c40ac|       111111111111111|               d9def48|      abe46783b919d312|               1111111|       d9def48277ca1f2|       111111111111111|       11f9765a24c40ac|      abe46783b919d312|
               d9df4c|                cede62|               d9def48|      b805f22782e1daa4|                     0|       d9def48277ca1f2|                cede62|                d9df4c|      b805f22782e1daa4|
       2fee91eb4a0588|        2d82d82d82d82e|               d9df4cb|      6f15aab46ebe4f18|                2d82d8|       d9df4cb805f2278|        2d82d82d82d82d|        2fee91eb4a0588|      6f15aab46ebe4f18|
               d9df51|                cede62|               d9df4cb|      2c817bc760df41f0|                     0|       d9df4cb805f2278|                cede62|                d9df51|      2c817bc760df41f0|
        6d8f047fe9483|         6806806806807|               d9df512|       9fb89f3541e5ac4|                 68068|       d9df512c817bc76|         6806806806806|         6d8f047fe9483|       9fb89f3541e5ac4|
               d9df51|                cede62|               d9df512|      661b1dced578792c|                     0|       d9df512c817bc76|                cede62|                d9df51|      661b1dced578792c|
         db1e0902bf8b|          d00d00d00d00|               d9df516|      a12a4b8288c80900|                  d00d|       d9df51661b1dced|          d00d00d00d00|          db1e0902bf8b|      a12a4b8288c80900|
               d9df51|                cede62|               d9df516|      6903d8696a6218ba|                     0|       d9df51661b1dced|                cede62|                d9df51|      6903d8696a6218ba|
          13ac805a643|                     1|               d9df516|      7da06f196f5d5961|                  171d|       d9df516903d8697|          171de3a556c7|           13ac805a643|      7da06f196f5d5961|
               d9df51|                cede62|               d9df516|      6929753adfc677ce|                     0|       d9df516903d8697|                cede62|                d9df51|      6929753adfc677ce|
          26f445de61f|           24fc9f6ef14|               d9df516|      22ad20582f8ea7ea|                   24f|       d9df516929753ae|           24fc9f6ef13|           26f445de61f|      22ad20582f8ea7ea|
               d9df51|                cede62|               d9df516|      692b5b88ae82ec9c|                     0|       d9df516929753ae|                cede62|                d9df51|      692b5b88ae82ec9c|
           38a91fcf19|            35cc8acfeb|               d9df516|      a642610e4941140e|                    35|       d9df51692b5b88b|            35cc8acfea|            38a91fcf19|      a642610e4941140e|
               d9df51|                cede62|               d9df516|      692b74183f732f36|                     0|       d9df51692b5b88b|                cede62|                d9df51|      692b74183f732f36|
            4b8c2a697|             47bb63bff|               d9df516|      4b9cc108cdc46cf8|                     4|       d9df51692b74184|             47bb63bfe|             4b8c2a697|      4b9cc108cdc46cf8|
               d9df51|                cede62|               d9df516|      692b7555cd2f8c88|                     0|       d9df51692b74184|                cede62|                d9df51|      692b7555cd2f8c88|
             5cfb47e3|              5849184e|               d9df516|      9af9a6413361ba56|                     0|       d9df51692b7555d|              5849184e|              5cfb47e3|      9af9a6413361ba56|
               d9df51|                cede62|               d9df516|      692b7565d70b539a|                     0|       d9df51692b7555d|                cede62|                d9df51|      692b7565d70b539a|
              6a43bfd|               64e5d2a|               d9df516|      d24468e32ac9f442|                     0|       d9df51692b7565d|               64e5d2a|               6a43bfd|      d24468e32ac9f442|
               d9df51|                cede62|               d9df516|      692b7566a5e9b59a|                     0|       d9df51692b7565d|                cede62|                d9df51|      692b7566a5e9b59a|
               715955|                6b9fd0|               d9df516|      9bff7a1831f00c20|                     0|       d9df51692b7566a|                6b9fd0|                715955|      9bff7a1831f00c20|
               d9df51|                cede62|               d9df516|      692b7566b06b0094|                     0|       d9df51692b7566a|                cede62|                d9df51|      692b7566b06b0094|
                71595|                 6b9fd|               d9df516|      59bff7a18325babf|                     0|       d9df51692b7566b|                 6b9fd|                 71595|      59bff7a18325babf|
               d9df51|                cede62|               d9df516|      692b7566b139def6|                     0|       d9df51692b7566b|                cede62|                d9df51|      692b7566b139def6|
                 6aae|                  654b|               d9df516|      4ef8b57f50ce8859|                     0|       d9df51692b7566b|                  654b|                  6aae|      4ef8b57f50ce8859|
               d9df51|                cede62|               d9df516|      692b7566b139def6|                     0|       d9df51692b7566b|                cede62|                d9df51|      692b7566b139def6|
                  5ed|                   5a1|               d9df516|      a620940fd9fd704b|                     0|       d9df51692b7566b|                   5a1|                   5ed|      a620940fd9fd704b|
               d9df51|                cede62|               d9df516|      692b7566b139def6|                     0|       d9df51692b7566b|                cede62|                d9df51|      692b7566b139def6|
                   50|                    4c|               d9df516|       ae4c2b38e6da7c4|                     0|       d9df51692b7566b|                    4c|                    50|       ae4c2b38e6da7c4|
               d9df51|                cede62|               d9df516|      692b7566b139def6|                     0|       d9df51692b7566b|                cede62|                d9df51|      692b7566b139def6|
                    4|                     4|               d9df516|      3677d45a4add59ac|                     0|       d9df51692b7566b|                     4|                     4|      3677d45a4add59ac|
               d9df51|                cede62|               d9df516|      692b7566b139def6|                     0|       d9df51692b7566b|                cede62|                d9df51|      692b7566b139def6|

         input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|
      ---------------+---------------+---------------+---------------+---------------+
       0.050505049527|   -11.06623662|     1.05180217|    21.66864677|121180.38792667|



PASS case
               absIn0|                absIn1|                 in0Hi|                 in0Lo|                 in1Hi|                  in_0|                  in_1|          ptroutbitshi|          ptroutbitslo|
---------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+
                   ce|                     0|                     0|      c000000000000000|              40000000|                   33b|      4000000000000000|                    ce|      c000000000000000|
                    0|                     0|                     0|             29c486696|                     0|                   33b|                cede62|                     0|             29c486696|
                    a|                     0|                     0|      8000000000000000|              40000000|                    2a|      4000000000000000|                     a|      8000000000000000|
                    0|                     0|                     0|              21f07c14|                     0|                    2a|                cede62|                     0|              21f07c14|
                    0|                     0|                     0|      2aaaaaaaaaaaaaaa|              15555555|                     2|      1555555555555555|                     0|      2aaaaaaaaaaaaaaa|
                    0|                     0|                     0|               19dbcc4|                     0|                     2|                cede62|                     0|               19dbcc4|

	      input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|            ULP|
           ---------------+---------------+---------------+---------------+---------------+
            0.050505049527|     1.05177820|     1.05180217|   -92.40404654|     0.23977156|
```

Reply:
@ShriramShastry I cant see this in the new code, has this assignment issue now resolved ? Sometimes re-writing the function is enough to avoid the CC bug.

Reply:
I'm saying `absin1 = 0` is removed, and I've provided an explanation if '`absin1 = 0`' is removed.
Though I'm not sure where this '`absin1 = in0lo absin0? 1L: 0L`' '`L`' is seen by @cujomalainey and why we're talking about compilers (I can't comment because the present code is '`absin1 = in0lo absin0? 1: 0;`'.

Reply:
@ShriramShastry  its [here](https://github.com/thesofproject/sof/pull/6655/files#diff-aec514d2a73ba965f4c6794ce24ee85f327bf2ecc274bb38f9ff5f4e94db53c4R122) line 122. You do a write without a read which just means its not needed. Compiler is likely throwing the instruction away anyways.

Reply:
OMG, I didn't realize this spurious `absin1 = 0` assignment was still there, good catch @cujomalainey , thank you.

@ShriramShastry the reason we're talking about compilers is because every **correct** compiler will discard this pointless line that does nothing. When I tried both the Xtensa XCC and the Zephyr SDK 0.15.2 toolchain above they both correctly discarded this pointless line of code. If you have found a compiler/configuration that does not discard this pointless line then you have found a very, very serious compiler issue that we MUST know about. So which exact toolchain does NOT discard this line? 

In other words, please share the toolchain and command line that creates a different `exp_fnc.c.obj` file when you replace this line by `// absin1 = 0`. No explanation please, only the toolchain and command line that produce a different `exp_fnc.c.obj`

Reply:
done, Thank you